src/ajax.js

```diff
@@ -444,80 +444,80 @@ jQuery.extend( {
 			// Default abort message
 			strAbort = "canceled",
 
-			// Fake xhr
-			jqXHR = {
-				readyState: 0,
+			// Fake xhr (ref the true Deferred instance above)
+			jqXHR = deferred;
```

Comment:
I understand what he is trying to say, and you are restating that jQuery is trying to protect ppl from stepping on their own foot, or screw someone else's code when loaded together. I, of course, understand and appreciate that concept and I will probably want to do the same whenever I can. The question is why and to what extend. Like if you have C pointers, you can kill your own stack, or in the worst case all the other programs if your OS doesn't protect it. In browser, your tab JS engine is the OS, and JS is C, jQuery is probably some relevant lib. Now, you see what I'm asking right? Why would a lib want to *protect* when it is only providing APIs and unit routines?

But again, the argument is based on my way of giving both Deferred and promise the class $.Deferred when there is no other options in the jQuery code base that I can use to achieve my goal of giving jqXHR a type. So based on my implementation, I ask @gibson042 to give me an example of a hackable situation, not examples of programmers stepping on their own foot. Like if every Chef holding a knife kills him/herself then we are in real trouble. Our purpose and caution when using a tool would be the only guard there, not the tool.

The JS language was built without protection in the first place, I think clear typing is far more important than passing plain *protected* objects around, especially when it comes to Promise. Promise is an important contract, the object that has that contract should have that type. Give me one counter example in good OO standard that's against it then I'll shut up.

A language itself should protect enough and leave room for the programmers, a library can only do so much about additional APIs. And in APIs we need to be clear about returns. This is why we want to hook up *your* APIs in the first place and why we are all in the Open Source community. We communicate clearly to each other, and provide reusable routines with meaningful interfaces (both in fn signature and return types). Not hidden caveat or some hacked policy when there is clearly a better way of interfacing. 

But again, returning a chopped object with minimum api is OK if you merge it with a type first in jQuery, so I do hope you consider giving all the Deferred and promise a proper type in v3. If $.Deferred is considered then having exposed `resolve` and `reject` shouldn't be the reason to drop it, since they are loved by our testers & devs when faking ajax calls in a server-less UI dev setup.